### ISSUE TITLE
Allow Configuring Git Protocol Used When Cloning for `flutter pub get`

### ISSUE THREAD

**Issue by zicklag** [flutter/flutter#81646]
## Use case

I am behind a corporate firewall that blocks all access to the internet except through an HTTP proxy server. When using flutter, and running `fluter pub get` I run into an issue where the git URLs that it is trying to clone, which use the `git://` protocol are blocked by my firewall. Switching the URL from `git://` to `https://` fixes the issue, but I could find no way to make flutter use the different protocol.

My current workaround is to create a small shell script that I name `git` and put in place of the real git executable that switches the git clone URL in the context of those specific flutter clone commands:

```sh
#/bin/bash
# Hack to convert git:// clone urls to http:// clone urls for the flutter dependency downloads
if [ "$FLUTTER_SHIM" = "true" -a "$2" = "--mirror" ]; then
    url=$(echo $3 | sed 's/git:\/\//https:\/\//')
    /usr/bin/git clone --mirror $url $4
else
    /usr/bin/git $@
fi
```

## Proposal

Flutter could provide an environment variable and a commandline flag to configure the git protocol used to clone when doing a `flutter pub get`. This flag could be set to either `git` or `https`.

---

Let me know if there is something I missed or a better way to accomplish this.

**Comment by jmagman** [comment]
1. What's the output of `flutter pub get -v` without your git hack?
2. What's the output of `flutter doctor -v`?

**Comment by zicklag** [comment]
Well, now I'm confused because it's just working without the git hack. :thinking: I'm not sure if it's because I've updated flutter since and it changed the way it did the git clone or if something in my environment changed. Or maybe it is that cloning it with https:// the first time made it use HTTP for all of the other times since and I need to clear some cache directory?

At this point maybe we can just close this because I can't reproduce and it seems to have been fixed.

Here's my flutter version:

```
Flutter 2.3.0-2.0.pre.44 • channel master • https://github.com/flutter/flutter.git
Framework • revision 14d5eab43c (12 days ago) • 2021-05-05 21:51:53 -0700
Engine • revision 378e4dbc41
Tools • Dart 2.14.0 (build 2.14.0-48.0.dev)
```

I'm good closing this if you don't think we should investigate further.

**Comment by jmagman** [comment]
@zicklag If you see this again, try to pinpoint where the git url is coming from using the `--verbose` flag.  If it's a particular package, you may be able to work around it by specifying it as a git package with a `https` URL in your pubspec: https://dart.dev/tools/pub/dependencies#git-packages.

If it seems to be coming from `pub`, you can file an issue in the pub repo https://github.com/dart-lang/pub.  And if it seems to be coming from Flutter (like checking if there's a newer version of Flutter) or if you're not sure, feel free to file a new issue, include the verbose logs, and we'll take a look!

**Comment by zicklag** [comment]
Awesome, thank you for the help!
